File: dp/number_distinct_subsequences.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function template for C++

class Solution {
 public:
  int distinctSubsequences(string s) {
    // Your code goes here
    int n = s.size();
    const int mod = 1e9 + 7;
    vector<int> dp(n + 1);
    dp[0] = 1;
    unordered_map<int, int> last_index;
    for (int i = 0; i < n; ++i) {
      dp[i + 1] = (dp[i] * 2) % mod;
      if (last_index.find(s[i]) != last_index.end()) {
        dp[i + 1] = (dp[i + 1] + mod - dp[last_index[s[i]]]) % mod;
      }
      last_index[s[i]] = i;
    }
    return dp[n];
  }
};

//{ Driver Code Starts.

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(NULL);
  cout.tie(NULL);

  int t;
  cin >> t;
  while (t--) {
    string s;
    cin >> s;

    Solution ob;
    cout << ob.distinctSubsequences(s) << "\n";
  }

  return 0;
}
// } Driver Code Ends

File: dp/number_distinct_subsequences.py
# User function Template for python3


class Solution:
    def distinctSubsequences(self, S):
        # code here
        M = int(1e9 + 7)
        n = len(S)
        last_index = {}
        dp = [0] * (n + 1)
        dp[0] = 1
        for i in range(n):
            if S[i] not in last_index:
                dp[i + 1] = dp[i] * 2 % M
            else:
                dp[i + 1] = (dp[i] * 2 - dp[last_index[S[i]]]) % M
            last_index[S[i]] = i

        return dp[n]


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        s = input()

        ob = Solution()
        answer = ob.distinctSubsequences(s)
        print(answer)

# } Driver Code Ends
